csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/FilePathFilter.cs

```diff
@@ -0,0 +1,96 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Security.Cryptography.X509Certificates;
+using System.Text.RegularExpressions;
+using Semmle.Util;
+using Semmle.Util.Logging;
+
+namespace Semmle.Extraction.CSharp.DependencyFetching
+{
+    public class FilePathFilter
+    {
+        private readonly string rootFolder;
+        private readonly IProgressMonitor progressMonitor;
+
+        public FilePathFilter(DirectoryInfo sourceDir, IProgressMonitor progressMonitor)
+        {
+            rootFolder = FileUtils.ConvertToUnix(sourceDir.FullName.ToLowerInvariant());
+            this.progressMonitor = progressMonitor;
+        }
+
+        private class FileInclusion(string path, bool include)
+        {
+            public string Path { get; } = path;
+
+            public bool Include { get; set; } = include;
+        }
+
+        private record class PathFilter(Regex Regex, bool Include);
+
+        public IEnumerable<FileInfo> Filter(IEnumerable<FileInfo> files)
+        {
+            var filters = (Environment.GetEnvironmentVariable("LGTM_INDEX_FILTERS") ?? string.Empty).Split(new[] { '\r', '\n' }, StringSplitOptions.RemoveEmptyEntries);
+            if (filters.Length == 0)
+            {
+                return files;
+            }
+
+            var pathFilters = new List<PathFilter>();
+
+            foreach (var filter in filters)
+            {
+                bool include;
+                string filterText;
+                if (filter.StartsWith("include:"))
+                {
+                    include = true;
+                    filterText = filter.Substring("include:".Length);
+                }
+                else if (filter.StartsWith("exclude:"))
+                {
+                    include = false;
+                    filterText = filter.Substring("exclude:".Length);
+                }
+                else
+                {
+                    progressMonitor.Log(Severity.Info, $"Invalid filter: {filter}");
+                    continue;
+                }
+
+                var regex = new FilePattern(filterText).RegexPattern;
+                progressMonitor.Log(Severity.Info, $"Filtering {(include ? "in" : "out")} files matching '{regex}'. Original glob filter: '{filter}'");
+                pathFilters.Add(new PathFilter(new Regex(regex, RegexOptions.IgnoreCase | RegexOptions.Compiled | RegexOptions.Singleline), include));
+            }
+
+            var unfilteredResult = files.Select(f =>
+                new
+                {
+                    FileInfo = f,
+                    FileInclusion = new FileInclusion(
+                        FileUtils.ConvertToUnix(f.FullName.ToLowerInvariant()).Replace(rootFolder, string.Empty).TrimStart('/'),
+                        pathFilters.All(f => !f.Include))
```

Comment:
Perhaps compute this up front, so we don't have to do it for all elements in `files`.

Reply:
That way we can also give it a describing name, e.g. `var includeByDefault = pathFilters.All(f => !f.Include)`.

Reply:
Good point, thanks.